Monika After Story/game/mod_assets/connect4/connect4.rpy

```diff
@@ -0,0 +1,655 @@
+                winning_cels.extend(self.get_winning_line_up(self.PLAYER, line_up))
+            for line_up in self.get_horizontal():
+                winning_cels.extend(self.get_winning_line_up(self.PLAYER, line_up))
+            for line_up in self.get_diagonal_ltr():
+                winning_cels.extend(self.get_winning_line_up(self.PLAYER, line_up))
+            for line_up in self.get_diagonal_rtl():
+                winning_cels.extend(self.get_winning_line_up(self.PLAYER, line_up))
+            return winning_cels
+    
+        def ai_move(self):
+            """
+            Tells the AI to move
+            """
+            # first we get the cels where someone can win
+            ai_winning_try_moves = list(self.get_ai_winning_cels())
+            player_winning_try_moves = list(self.get_player_winning_cels())
+            
+            while not self.turn_move_success:
+                renpy.pause(1.5)
+                
+                # the AI checks where it can win first. if found, makes a move there to win
+                while ai_winning_try_moves and not self.turn_move_success:
+                    try_fill = random.choice(ai_winning_try_moves)
+                    is_bottom_good = try_fill[0] == '5' or self.board.get(self.g_k(int(try_fill[0]) + 1, int(try_fill[1])), None) is not None
+                    if is_bottom_good:
+                        self.col_fill(try_fill[1], self.AI)
+                    else:
+                        ai_winning_try_moves.remove(try_fill)
+                
+                # if there are no winning move, check for tiles where player can win to cancel them from winning
+                while player_winning_try_moves and not self.turn_move_success:
+                    try_fill = random.choice(player_winning_try_moves)
+                    is_bottom_good = try_fill[0] == '5' or self.board.get(self.g_k(int(try_fill[0]) + 1, int(try_fill[1])), None) is not None
+                    if is_bottom_good:
+                        self.col_fill(try_fill[1], self.AI)
+                    else:
+                        player_winning_try_moves.remove(try_fill)
+                        
+                # makes a random move if there are no winning tiles lol
+                if not self.turn_move_success:
+                    self.col_fill(renpy.random.choice(list({i for i in range(self.CEL_COL)} - self.full_col)), self.AI)
+        
+    
+    
+label game_connect4:
+    m "Let's start!"
+    
+    label start_connect4:
+        pass
+        
+    window hide None
+    show monika 1eua at t21
+    python:
+        quick_menu = False
+        
+        connect4_displayable_obj = MASConnect4Displayable()
+        connect4_displayable_obj.show()
+        results = connect4_displayable_obj.game_loop()
+        connect4_displayable_obj.hide()
+        
+        quick_menu = True
+        
+        winner, ending = results
+        
+    label .connect4_end:
+        pass
```

Comment:
probably don't need this label, i don't see it used anywhere else.

Reply:
I used to jump to that label from the console. Beyond that, it has no other use.